figma/lib/fetch_components.ts

```diff
@@ -0,0 +1,246 @@
+  }
+  const type: PropTypes = []
+  const nonBoolValues = stripBooleanValues(prop.values)
+  const simpleType = nonBoolValues.length === 1
+  const bool = prop.values.length !== nonBoolValues.length && !simpleType
+  const values = nonBoolValues.filter((v) => v !== DEFAULT_LABEL)
+  const required = presentInAllVariants && values.length === nonBoolValues.length && !bool
+  if (bool) type.push('boolean')
+  return { ...prop, values, required, type }
+}
+
+function processVariants(output: RawOutput<MergedComponentType>): Output {
+  const res: Output = {}
+
+  Object.keys(output).forEach((componentKey) => {
+    const states: string[] = []
+    const props: Props = {}
+    const counters: Record<string, number> = {}
+    const { variants, ...component_ } = output[componentKey]
+    const variantKeys = Object.keys(variants)
+    variantKeys.forEach((variantKey) => {
+      const variant = variants[variantKey]
+      const keyValues = variantNameToKeyValues(variant.name)
+      if (!keyValues.length) return
+      keyValues.forEach(({ key, value }) => {
+        if (key.match(/property\s?[0-9]+/gi)) return
+        if (isStateKey(key)) {
+          if (!states.includes(value)) states.push(value)
+          return
+        }
+        if (!props[key]) props[key] = { name: key, values: [], required: true }
+        if (!props[key].values.includes(value)) props[key].values.push(value)
+        counters[key] = counters[key] ?? 0
+        counters[key]++
+      })
+    })
+    objToArr(props).forEach((prop) => {
+      props[prop.name] = processPropTypes(prop, counters[prop.name] === variantKeys.length)
+    })
+    const component = { ...component_, states: cleanStateValues(states), props }
+    res[componentKey] = component
+  })
+
+  return res
+}
+
+function isThemeKey(key: string): boolean {
+  return (THEME_KEYS as string[]).includes(key)
+}
+
+function getNameAndPath(origName: string) {
+  const nameSections = origName
+    .split(/\//g)
+    .map((s) => s.trim())
+    .filter(Boolean)
+
+  const path = nameSections.filter((section) => ![...THEME_KEYS, 'global'].includes(section)).join('/')
+  const name = camelize(nameSections[0], true)
+  const lastSection = nameSections[nameSections.length - 1]
+  const theme = isThemeKey(lastSection) ? lastSection : undefined
+  return { name, path, theme }
+}
+
+export default async function fetchComponents(fileId: string) {
+  const res = await Figma.getFile(fileId)
+  const fileName = res.name
+  const lastModified = res.lastModified
+  const topLevelCanvas = getFileTopLevelChildren(res.document)
+  const output: RawOutput = {}
+  let frameName = ''
+  let frameId = ''
+  topLevelCanvas.forEach((topLevel) =>
+    getWalk(extractors)((data, parent) => {
+      const [id, type, node] = data
+      const canvasName = topLevel.name
+      if (type === 'FRAME') {
+        frameName = node.name
+        frameId = node.id
+        return
+      }
+      if (type === 'COMPONENT_SET') {
```

Comment:
I almost never use else unless I am absolutely forced to, or it is to complicated to do so, basically it's a guard pattern so you exit the code and branching does not even need to be checked. It's a pretty minor optimisation (esp. that probably webpack will do it for me anyways), but I try to avoid nesting code as much as possible